### PR TITLE
Add lockfiles and caching for remote dependencies

### LINKED ISSUE #9: Add support for lock files
Deno provides the ability to use lock files: https://deno.land/manual/linking_to_external_code/integrity_checking

It would be great to add a feature to the `deno_binary` rule to allow this. Something like:

```skylark
deno_binary(
    name = "my_script",
    main = "my_script.ts",
    lock_file = "lock.json",
)
```

Additionally, you should be able to update that lock file as you would with a normal invocation with the `--lock-write` flag. It might be good for the binary to create another target with a `_lock_write` suffix to do this.

```shell
bazel run :my_script_lock_write
```

### PR DESCRIPTION
**Pull request by stabai**
fixes #9 

To use lockfiles, users have to manually specify a lockfile. But the caching is automatic (can optionally turn off enforcement that cache is exhaustive).

This means that:
* Build will fail if remote dependencies don't match lock
* Build will fail if remote dependency caching fails
* Binary will fail to run if cache is incomplete
* Binary will fail to run if cache does not match lock
* No remote dependencies will ever be fetched at runtime (because cache is exhaustive and part of build tree)

### CODE REVIEW

**Review by alexeagle — COMMENTED** [review]
This is a little surprising. Usually Bazel is involved in download and cache external dependencies, then they are supplied as inputs to actions. Do you have some design trade off for doing it this way?

**Comment by stabai** [comment]
> This is a little surprising. Usually Bazel is involved in download and cache external dependencies, then they are supplied as inputs to actions. Do you have some design trade off for doing it this way?

Yeah, this is analogous to the way that one might depend on NPM packages, except that Deno forgoes the package.json file by importing remote files directly inside scripts: https://github.com/aspect-build/rules_deno/blob/3189cee8a66b08358799683b702fe32c92ed6f13/e2e/workspace/deno_binary/main.ts#L1
(this is something that these rules already support, but they're currently doing all that magic at runtime and ignoring them at build time)

I think the most Bazel-ey way to do this would probably be to have a `deno_imports` repository rule that installs all of the dependencies for a given script, and creates a repository for all such remote includes like `@deno//deno.land/std@0.153.0/fmt:colors.ts` (like how rules_nodejs uses `npm_install` to create `@npm//foo`). Then, the `deno_binary` rules could add those repository targets to `deps`. I didn't do that because it adds a lot of extra management steps that I'm not sure provide a ton of value. Deno does all of this automatically under the hood, so the lock file is completely optional and only used if you (rightly) want to ensure remote package integrity. While it's weird for the Bazel workspace to have external dependencies not declared in the workspace file, it also feels weird to break Deno's dependency management capabilities so that we can put them back together in a more awkward way. 🤷‍♂️

As this PR is written, these remote dependencies are only deterministic at the target level (falling short of the Bazel norm of making remote decencies deterministic at the workspace level). But it also allows the writing and management of Deno scripts to be about as easy as it would be without Bazel. This felt like a good trade-off to me, but open to your thoughts.

**Comment by stabai** [comment]
I guess I'm also not sure how involved Bazel should be in downloading and caching dependencies from another build tool. Is it that we want Bazel to have a repository rule to fetch all of those dependencies at the proper stage? Or that we want it to be aware of every particular dependency?

If it's just the former, we could resolve this by having a repository rule fetch all of the remote dependencies for the script and providing a repository where all of those dependencies could be included together:

```starlark
deno_remote_deps(
    name = "foobar_deps",
    main = "//foo/bar:main.ts",
    lock_json = "//foo/bar:lock.json",
)
```

Then the target for the script itself could use the `foobar_deps` repository created by that:

```starlark
# Option 1
deno_binary(
    name = "bar",
    main = "main.ts",
    deps = ["@foobar_deps//:remote_deps"],
)
```
I think this would be good, and not make things overly difficult. The workspace initialization would be responsible for fetching remote dependencies, as is typically the case for Bazel.

The difficulty is if `@foobar_deps//:remote_deps` is too broad, and we want targets to list all of those dependencies individually:

```starlark
# Option 2
deno_binary(
    name = "bar",
    main = "main.ts",
    deps = [
        "@foobar_deps//deno.land/std@0.153.0/fmt:colors.ts",
        "@foobar_deps//deno.land/std@0.153.0/bytes:mod.ts",
        "@foobar_deps//deno.land/std@0.153.0/fs:mod.ts",
        // etc for every remote import
    ],
)
```
This seems like unnecessary extra work to me, since we always want to include everything it found as a dependency of that script.

WDYT? Is Option 1 a good way forward?

**Comment by alexeagle** [comment]
sorry I didn't make time to dig into this before vacation, maybe @gregmagolan can give some suggestions as the primary author of rules_js dependency fetches

**Comment by gregmagolan** [comment]
I should have some time to TAL tomorrow

File: deno/private/binary.bzl [hunk 1]
```diff
@@ -17,32 +17,57 @@ def _deno_binary_impl(ctx):
...
+    ctx.actions.run(
+        executable = deno_tools[0].path,
+        arguments = ["cache", ctx.file.main.path] + build_flags,
+        tools = deno_tools,
+        env = {
+            "DENO_DIR": cache_folder.path,
+        },
+        mnemonic = "DenoCache",
+        inputs = infiles,
+        outputs = [cache_folder],
```
Comment (gregmagolan) [inline deno/private/binary.bzl:51]: probably want to add the `requires-network` execution requirement here... although remote execution environment that are network sandboxed don't seem to honour it it seems

**Comment by gregmagolan** [comment]
This is a really interesting and unique shaped cache being a build output. I see a few downsides to this approach that I'll enumerate:

1. each deno_binary will get its own cache output that can contain duplicate dependencies and result in multiple downloads if the same dependency is used in multiple targets

2. different platforms get their own output tree so for a single deno_binary you may get multiple copies in the output tree for different platforms which will result in multiple downloads

3. the build cannot be network isolated since build steps necessarily reach out to the network to fetch dependencies to build their cache outputs

4. the cache build output for any deno_binary is busted if any inputs change so a change to a comment in a source file that is an input to the deno_binary would result re-running the `deno cache` action and re-fetching the dependencies from the network

Parsing the lock file and created an external repository configuration that can lazy fetch the 3rd party deps when an action depends on them, similar to what we do in rules_js, would mitigate all of the downsides at the cost of some added complexity in the repository rule. It looks like the deno lock file is in JSON format so it would be easy to load in a repository rule now that [json.decode](https://github.com/aspect-build/rules_js/blob/86b7aaa3d4bcafaf30bbaf04fba2b7c1d770c078/npm/private/npm_import.bzl#L397) is built right into Starlark.

We've got some decent patterns for generating lazy fetched 3rd party dependencies in rules_js that could be borrowed for the lockfile generation and index repository and a repository for each dep itself (needed to make the fetching lazy). I think ultimately that is a better solution even if more work upfront. In particular it feels akward to re-fetch 3rd party deps when a source file changes.

**Comment by stabai** [comment]
@gregmagolan Sorry I haven't gotten time to loop back around to this. The feedback is super useful, thank you. Parsing the lock file in a repository rule to generate repositories for each dep makes sense. My only question is: since Deno does locks on a per-file basis, does that make it ideal for each file to be its own repository? My current thinking is to do the following:

Have a `repository_rule` (e.g. `deno_imports`) that will take a lock file as an input and:
1. Create an array to hold all of the generated targets for deps
2. Loop through each entry in the lock file
    a. Convert the fully-qualified filename into a valid repository name
        (e.g. `https://deno.land/std@0.153.0/fmt/colors.ts` => `deno_land-std_at_0.153.0-fmt-colors.ts`)
    b. Create an `http_file` rule containing the file
    c. Add `@deno_land/std_at_0.153.0_fmt_colors.ts//file` to the array of target deps
3. Convert the workspace-qualified lock filename into a valid repository name
    (e.g. `//my_app/main.lock` => `my_app-main.lock`)
5. Create a directory for a rollup of the deps for that the lock file
6. Write a WORKSPACE file and BUILD file to that directory
     a. Create a `filegroup` named `deps` that includes all of the deps from the array
7. Use `local_repository` to turn that folder into a repository

Update `deno_binary` to use the lock file:
1. Convert the workspace-qualified lock filename into a valid repository name
2. Add `@my_app-main.lock//:deps` to the deps when building

Does this seem sensible? I've never built this kind of rule before, so please let me know if this is not what you meant, or if you have any pointers to code you recommend as a guide. Thank you!

**Comment by alexeagle** [comment]
Hey @stabai we should figure out how to move this ahead, maybe find 30min to chat with Greg on a VC? https://calendly.com/gregmagolan/consult

**Comment by gregmagolan** [comment]
@stabai We migrated the home of rules_deno to GitLab since we needed some of our open source repositories on GitLab to serve as demonstration GitLab deployments of [Aspect Workflows](https://github.com/aspect-build/rules_deno/pull/aspect.build/workflows).

If you still want to keep this PR alive can you re-create it as GitLab MR where the new home of rules_deno? The repo is now at https://gitlab.com/aspect-build/rules_deno.

This PR migrated to an [MR](https://gitlab.com/aspect-build/rules_deno/-/merge_requests/14) but the author attribution was lost on the migration so I closed that one.

**Pull request closed without merging**